### PR TITLE
Fix for 774

### PR DESCRIPTION
**Pull request by sbrossie**

### CODE REVIEW

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java [hunk 1]
```diff
@@ -0,0 +1,628 @@
...
+        assertListenerStatus();
+
+        invoiceChecker.checkInvoice(account.getId(), 2, callContext, new ExpectedInvoiceItemCheck(new LocalDate(2017, 12, 1), new LocalDate(2018, 1, 1), InvoiceItemType.RECURRING, new BigDecimal("249.95")));
+
+        // Future pause the subscription
+        LocalDate effectivePauseDate = new LocalDate(2017, 12, 15);
+        entitlementApi.pause(baseEntitlement.getBundleId(), effectivePauseDate, ImmutableList.<PluginProperty>of(), callContext);
+
+        // Verify the next invoice based on the PAUSE event is correctly seen in the dryRun scenario
+        expectedInvoices.add(new ExpectedInvoiceItemCheck(new LocalDate(2017, 12, 15), new LocalDate(2018, 1, 1), InvoiceItemType.REPAIR_ADJ, new BigDecimal("-137.07")));
```
Comment (pierre) [inline beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java]: Would we see the same if we hadn't moved the clock? We probably wouldn't have a repair in that case?

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java [hunk 2]
```diff
@@ -0,0 +1,628 @@
...
+        assertListenerStatus();
+
+        // Move clock after PHASE event
+        busHandler.pushExpectedEvents(NextEvent.PHASE, NextEvent.INVOICE, NextEvent.INVOICE_PAYMENT, NextEvent.PAYMENT);
+        clock.addMonths(12);
+        assertListenerStatus();
+    }
+
+    @Test(groups = "slow")
+    public void testDryRunWithPendingCancelledSubscription() throws Exception {
```
Comment (pierre) [inline beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java]: No dry-run in this test?

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+        assertEquals(invoices.size(), 2);
+
+        final List<ExpectedInvoiceItemCheck> toBeChecked = ImmutableList.<ExpectedInvoiceItemCheck>of(
+                new ExpectedInvoiceItemCheck(new LocalDate(2017, 5, 1), new LocalDate(2017, 6, 1), InvoiceItemType.REPAIR_ADJ, new BigDecimal("-19.95")),
+                new ExpectedInvoiceItemCheck(new LocalDate(2017, 5, 1), new LocalDate(2017, 5, 1), InvoiceItemType.CBA_ADJ, new BigDecimal("19.95")));
+        invoiceChecker.checkInvoice(invoices.get(1).getId(), callContext, toBeChecked);
+    }
+
+    @Test(groups = "slow", description = "https://github.com/killbill/killbill/issues/783")
+    public void testIntegrationWithRecurringFreePlan() throws Exception {
```
Comment (pierre) [inline beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java]: No dry-run?

File: invoice/src/main/java/org/killbill/billing/invoice/notification/DefaultNextBillingDatePoster.java [hunk 4]
```diff
@@ -81,25 +94,28 @@ private void insertNextBillingFromTransactionInternal(final EntitySqlDaoWrapperF
...
                 nextBillingQueue.recordFutureNotificationFromTransaction(entitySqlDaoWrapperFactory.getHandle().getConnection(), futureNotificationTime,
-                                                                         new NextBillingDateNotificationKey(subscriptionId, targetDate, isDryRunForInvoiceNotification), internalCallContext.getUserToken(),
+                                                                         newNotificationEvent, internalCallContext.getUserToken(),
                                                                          internalCallContext.getAccountRecordId(), internalCallContext.getTenantRecordId());
-            } else if (log.isDebugEnabled()) {
-                log.debug("*********************   SKIPPING Queuing next billing date notification at {} for subscriptionId {} *******************", futureNotificationTime.toString(), subscriptionId.toString());
+            } else {
+                log.info("Updating next billing date notification event at {} for subscriptionId {}", futureNotificationTime.toString(), JOINER.join(subscriptionIds));
```
Comment (pierre) [inline invoice/src/main/java/org/killbill/billing/invoice/notification/DefaultNextBillingDatePoster.java]: Maybe `debug`?

File: api/src/main/java/org/killbill/billing/subscription/api/SubscriptionBaseInternalApi.java [hunk 5]
```diff
@@ -93,9 +93,7 @@ public SubscriptionBaseBundle createBundleForAccount(UUID accountId, String bund
 
     public void updateExternalKey(UUID bundleId, String newExternalKey, InternalCallContext context);
 
-    public Iterable<DateTime> getFutureNotificationsForAccount(InternalCallContext context);
-
-    public Map<UUID, DateTime> getNextFutureEventForSubscriptions(final SubscriptionBaseTransitionType eventType, final InternalCallContext internalCallContext);
+    public Map<UUID, DateTime> getNextFutureEventForSubscriptions(final List<SubscriptionBaseTransitionType> eventTypes, final InternalCallContext internalCallContext);
```
Comment (pierre) [inline api/src/main/java/org/killbill/billing/subscription/api/SubscriptionBaseInternalApi.java]: Is this still in use?

Comment (sbrossie) [reply api/src/main/java/org/killbill/billing/subscription/api/SubscriptionBaseInternalApi.java]: It's not, i thought of removing it, and somehow proceeded to not do it. The `SubscriptionDao#getFutureEventsForAccount` methods is also unused. Will remove both.

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java [hunk 6]
```diff
@@ -0,0 +1,628 @@
...
+    public void testDryRunWithNoTargetDate() throws Exception {
+        final int billingDay = 14;
+        final DateTime initialCreationDate = new DateTime(2015, 5, 15, 0, 0, 0, 0, testTimeZone);
+        // set clock to the initial start date
+        clock.setTime(initialCreationDate);
+
+        log.info("Beginning test with BCD of " + billingDay);
+        final Account account = createAccountWithNonOsgiPaymentMethod(getAccountData(billingDay));
+
+        int invoiceItemCount = 1;
```
[the remark this answers is not in the document]
Comment (sbrossie) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java]: Yeah, cut and paste from other tests, all our tests seem o use that... Will fix.

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java [hunk 7]
```diff
@@ -0,0 +1,628 @@
...
+        // One more time, this will verify that the upcoming invoice notification is found and the invoice is generated at the right date, with correct items
+        expectedInvoices.add(new ExpectedInvoiceItemCheck(new LocalDate(2015, 8, 14), new LocalDate(2015, 9, 14), InvoiceItemType.RECURRING, new BigDecimal("249.95")));
+        dryRunInvoice = invoiceUserApi.triggerInvoiceGeneration(account.getId(), null, DRY_RUN_UPCOMING_INVOICE_ARG, callContext);
+        invoiceChecker.checkInvoiceNoAudits(dryRunInvoice, callContext, expectedInvoices);
+    }
+
+    //
+    // More sophisticated test with two non aligned subscriptions that verifies the behavior of using invoice dryRun api with no date
+    // - The first subscription is an annual (SUBSCRIPTION aligned) whose billingDate is the first (we start on Jan 2nd to take into account the 30 days trial)
+    // - The second subscription is a monthly (ACCOUNT aligned) whose billingDate is the 14 (we start on Dec 15 to also take into account the 30 days trial)
```
[the remark this answers is not in the document]
Comment (sbrossie) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java]: Will fix.

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java (hunk 2 above)
Comment (sbrossie) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java]: Will add one to test a case of future cancelled subscription.

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java (hunk 3 above)
Comment (sbrossie) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java]: Will remove, this is a dup (defined in `TestIntegrationInvoice`)

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java (hunk 1 above)
Comment (sbrossie) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java]: Don't understand the question -- in particular we have not yet moved the clock to reach the PAUSE date.

File: invoice/src/main/java/org/killbill/billing/invoice/notification/DefaultNextBillingDatePoster.java (hunk 4 above)
Comment (sbrossie) [reply invoice/src/main/java/org/killbill/billing/invoice/notification/DefaultNextBillingDatePoster.java]: The one where we insert is `info`, should we be consistent between the two?

Comment (pierre) [reply invoice/src/main/java/org/killbill/billing/invoice/notification/DefaultNextBillingDatePoster.java]: Oh, I see - I was just afraid of flooding. We can always revisit later I suppose.

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java (hunk 1 above)
Comment (pierre) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/TestIntegrationDryRunInvoice.java]: The scenario is:

1. t=2017-11-1 CREATE
2. t=2017-12-1 PHASE (2017-12-1 -> 2018-1-1)
3. t=2017-12-1 PAUSE (future)
4. t=2017-12-15 PAUSE (2017-12-15 -> 2018-1-1 repair + CBA)

Now, if the scenario is:

1. t=2017-11-1 CREATE
2. t=2017-11-1 PAUSE (future)

and we trigger a dry-run: will we simply see a recurring item from 2017-12-1 to 2017-12-15 or do we see the expected invoice (recurring, repair, CBA)?

**Pull request merged**